lib/service/connect.go

```diff
@@ -62,8 +64,101 @@ import (
+	// by this instance's identity/join token. doing this before attempting allocation helps us avoid
+	// the situation where a user ends up creating a teleport instance that is running multiple services
+	// authorized by different join tokens, which poses challenges currently and may be a strictly
+	// invalid state in the future if we start enforcing access-control limitations on agents via
+	// their join tokens.
+	if role != types.RoleInstance && role != types.RoleAdmin && role != types.RoleAuth && process.getLocalAuth() == nil && !process.hostedPluginRoleExpected(role) {
+		state, err := process.storage.GetState(process.ExitContext(), role)
+		isNewRole := trace.IsNotFound(err)
+		if err != nil && !isNewRole {
+			return nil, trace.Wrap(err)
+		}
+
+		var isV16OrNewer bool
+		if !isNewRole {
+			if vs, ok := state.GetInitialLocalVersion(); ok {
+				v, err := semver.NewVersion(vs)
+				if err != nil {
+					// this field should always be valid semver
+					return nil, trace.Errorf("invalid initial local version %q in local agent state (this is a bug)", vs)
+				}
+
+				isV16OrNewer = v.Major >= 16
+			}
+		}
+
+		if isNewRole || isV16OrNewer {
+			conn := process.waitForInstanceConnector()
+			if conn == nil {
+				if process.ExitContext().Err() != nil {
+					return nil, trace.Errorf("connect attempt aborted (shutting down)")
+				}
+				return nil, trace.Errorf("got nil instance connector during connect attempt for role %q (this is a bug)", role)
+			}
+
+			// new/post-v16 system roles must be represented on the instance cert.
+			if !conn.ClientIdentity.HasSystemRole(role) {
+				process.logger.ErrorContext(process.ExitContext(), "refusing to start with system role not authorized by initial join token. consider creating a new join token and deleting the teleport state directory to reset this agent's identity.", "role", role, "authorized_roles", conn.ClientIdentity.SystemRoles)
+				return nil, trace.Errorf("role %q is not in set of authorized roles: %+v (consider creating a new join token and deleting the teleport state directory to reset this agent's identity)", role, conn.ClientIdentity.SystemRoles)
+			}
+		}
+	}
+
+	// if we're setting up the instance connector here its important that we check for system role mismatch issues. for system role
+	// mismatches introduced in earlier versions we need to attempt to grandfather them into the instance cert via assertions. for
+	// newly added roles we want to reject the addition explicitly. doing this helps us avoid the situation where a user ends up creating
+	// a teleport instance that is running multiple services authorized by different join tokens, which poses challenges currently
+	// and may be a strictly invalid state in the future if we start enforcing access-control limitations on agents via their join tokens.
+	if role == types.RoleInstance && process.getLocalAuth() == nil {
+		identity, err := process.GetIdentity(role)
```

Comment:
I just want to make sure I understand this right:

1. the check above is for non "instance" roles, i.e. `RoleDiscovery`, `RoleDatabase` etc.). It needs to wait for the "instance" connector to be broadcast.
2. this check is for the "instance" role registration itself, so we look to see if we already have an instance identity stored
3. if we do have an instance identity already, assert that we either have all the instance roles in that identity OR that the identity is older than v16
4. if we *dont* have an identity already, then this is first time connect and we will write the teleport version in "InitialLocalVersion"
5.  finally, the instance connector is made available to 1. and we assert again that the role is present in the instance identity